py/repl.c

```diff
@@ -143,6 +144,93 @@ bool mp_repl_continue_with_input(const char *input) {
     return false;
 }
 
+STATIC bool test_qstr(mp_obj_t obj, qstr name) {
+    if (obj) {
+        // try object member
+        mp_obj_t dest[2];
+        mp_load_method_protected(obj, name, dest, true);
+        return dest[0] != MP_OBJ_NULL;
+    } else {
+        // try builtin module
+        return mp_map_lookup((mp_map_t *)&mp_builtin_module_map,
+            MP_OBJ_NEW_QSTR(name), MP_MAP_LOOKUP);
+    }
+}
+
+STATIC const char *find_completions(const char *s_start, size_t s_len,
+    mp_obj_t obj, size_t *match_len, qstr *q_first, qstr *q_last) {
+
+    const char *match_str = NULL;
+    *match_len = 0;
+    *q_first = *q_last = 0;
+    size_t nqstr = QSTR_TOTAL();
+    for (qstr q = MP_QSTR_ + 1; q < nqstr; ++q) {
+        size_t d_len;
+        const char *d_str = (const char *)qstr_data(q, &d_len);
+        if (s_len <= d_len && strncmp(s_start, d_str, s_len) == 0) {
+            if (test_qstr(obj, q)) {
+                // special case; filter out words that begin with underscore
+                // unless there's already a partial match
+                if (s_len == 0 && d_str[0] == '_') {
```

Comment:
To make this more efficient, I think this test can be moved up 2 levels, to between the `qstr_data()` line and the `if (... strncmp)` line.

Reply:
@tyomitch I can do this during merge if that makes it easier for you.

Reply:
Sorry, been away from my dev PC for a few days :)

I've updated the patch, at the same time also addressing @stinos feedback by using original PR IDs instead of commit IDs for reference.